### PR TITLE
Catch exit in anayze

### PR DESCRIPTION
**Pull request by lukaszsamson**
Fixes https://github.com/JakeBecker/elixir-ls/issues/98

### CODE REVIEW

File: apps/language_server/lib/language_server/dialyzer/analyzer.ex [hunk 1]
```diff
@@ -57,6 +57,10 @@ defmodule ElixirLS.LanguageServer.Dialyzer.Analyzer do
 
     parent = self()
 
+    # dialyzer_analysis_callgraph.start exits on failure
+    # need to unset :trap_exit when analysis finises
+    Process.flag(:trap_exit, true)
+
     pid =
       spawn_link(fn ->
```
Comment (michalmuskala) [inline apps/language_server/lib/language_server/dialyzer/analyzer.ex]: The alternative to the `trap_exit` would be to use `spawn_monitor` here and receive on the monitor message.

**Comment by lukaszsamson** [comment]
Ping @JakeBecker, can we get this merged?
I'm still occasionally experiencing this crash on recent elixir/otp/ls

**Comment by JakeBecker** [comment]
Sorry for the wait. I'm hesitant to trap exits as it makes processes behave in a way I find more complex to think about. I think I have a better way of handling it:

Up till now, the reason that the `analyze` is triggered with a GenServer `call` instead of `cast` is just so that a build doesn't happen concurrently (since the build can delete beam files while the Dialyzer server tries to stale-check them). There are other places where the global state changes caused by builds are a problem too, and we added a global "build lock" that processes can grab to avoid this sort of thing.

I'm turning "analyze" into a GenServer cast, which allows the main language server to continue servicing requests. The Dialyzer server will grab the build lock while it stale-checks the beam files so it only blocks builds, not the main language server. This way, an exit in the Dialyzer server won't kill the main language server and the supervisor can restart the Dialyzer server.

New PR: https://github.com/JakeBecker/elixir-ls/pull/159

**Comment by lukaszsamson** [comment]
@JakeBecker I like your approach. Mine was the smallest change to make it more stable.

**Pull request closed without merging**